data/src/main/java/com/microsoft/azure/kusto/data/auth/endpoints/KustoTrustedEndpoints.java

```diff
@@ -0,0 +1,144 @@
+        overrideMatcher = matcher;
+    }
+
+    /**
+     * Throw an exception if the endpoint specified is not trusted.
+     *
+     * @param uri - Kusto endpoint
+     * @param loginEndpoint The login endpoint to check against.
+     * @throws KustoClientInvalidConnectionStringException - Endpoint is not a trusted Kusto endpoint
+     */
+    public static void ValidateTrustedEndpoint(String uri, String loginEndpoint) throws KustoClientInvalidConnectionStringException {
+        try {
+            ValidateTrustedEndpoint(new URI(uri), loginEndpoint);
+        } catch (URISyntaxException ex) {
+            throw new KustoClientInvalidConnectionStringException(ex);
+        }
+    }
+
+    /**
+     * Throw an exception if the endpoint specified is not trusted.
+     *
+     * @param uri - Kusto endpoint
+     * @throws KustoClientInvalidConnectionStringException - Endpoint is not a trusted Kusto endpoint
+     */
+    public static void ValidateTrustedEndpoint(String uri) throws KustoClientInvalidConnectionStringException {
+        try {
+            ValidateTrustedEndpoint(new URI(uri), CloudInfo.retrieveCloudInfoForCluster(uri).getLoginEndpoint());
+        } catch (URISyntaxException | DataServiceException ex) {
+            throw new KustoClientInvalidConnectionStringException(ex);
+        }
+    }
+
+    /**
+     * Is the endpoint uri trusted?
+     *
+     * @param uri The endpoint to inspect.
+     * @param loginEndpoint The login endpoint to check against.
+     * @throws KustoClientInvalidConnectionStringException - Endpoint is not a trusted Kusto endpoint
+     */
+    public static void ValidateTrustedEndpoint(URI uri, String loginEndpoint) throws KustoClientInvalidConnectionStringException {
+        Ensure.argIsNotNull(uri, "uri");
+
+        // Check that target hostname is trusted and can accept security token
+        ValidateHostnameIsTrusted(uri.getHost(), loginEndpoint);
+    }
+
+    /**
+     * Adds the rules that determine if a hostname is a valid/trusted Kusto endpoint
+     * (extends existing rules).
+     *
+     * @param rules   - A set of rules
+     * @param replace - If true nullifies the last added rules
+     */
+    public synchronized static void AddTrustedHosts(List<MatchRule> rules, boolean replace) {
+        if (replace) {
+            additionalMatcher = null;
+        }
+
+        if (rules == null || rules.isEmpty()) {
+            return;
+        }
+
+        additionalMatcher = FastSuffixMatcher.Create(additionalMatcher, rules);
+    }
+
+    private static void ValidateHostnameIsTrusted(String hostname, String loginEndpoint) throws KustoClientInvalidConnectionStringException {
+        // The loopback is unconditionally allowed (since we trust ourselves)
+        if (CloudInfo.isLocalAddress(hostname)) {
+            return;
+        }
+
+        // Either check the override matcher OR the matcher:
+        if (overrideMatcher != null) {
+            if (overrideMatcher.test(hostname)) {
```

Comment:
Should the additional matcher still be on w/ the override policy?

Reply:
i dont think its critical - whats important about the override matcher is that if it succeeds it will be return a match